packages/nx/src/explain/config-explainer.ts

```diff
@@ -0,0 +1,23 @@
+type Markdown = string;
+
+/**
+ * A ConfigExplainer should be an object which has the same keys as the generic `Config` which it
+ * is explaining, but importantly, regardless of the optionality of the keys of the underlying
+ * `Config`, all keys are required to be implemented (we use the TS built in `Required<T>` to
+ * enforce that).
+ *
+ * The value of each entry in the ConfigExplainer instance should be an object containing an
+ * `nxDevLearnMorePath` which will be used to link out to the docs as the source of truth for
+ * the generic field description and functionality, and an `explainConfig()` method which will be
+ * given the user's actual config for that field in order to provide a specific explanation.
+ */
+export type ConfigExplainer<Config> = Required<{
```

Comment:
This is very complex. Let's figure out how to abstract this later on.

We should be able to have a function:

```ts
function explainConfig(configPath: string, configKey: string): { url: string, description: string} {
  if (isNxJson(configPath)) {
    return explainNxJson(configPath, configKey)
  } else {
    // .....  
  }
}
```

Reply:
I get this more now after seeing packages/nx/src/explain/config-explainers/nx-json.ts.

I'm not sure if a `ts` file is the best place for this content :thinking:

Reply:
As you can see, conditionals, loops and recursive functions are all required at various points in order to accurately explain configs in a human-readable way. Please can you give some examples of how that level of dynamic logic can be achieved outside of runtime code?

The editor tooling already picks up on the description of the fields for me: 
![image](https://user-images.githubusercontent.com/900523/166206994-b40dd01e-c908-425d-a33a-49f96d436b4b.png)

But in order to accurately explain the user's actual config (the focus of this PR), I believe code is needed